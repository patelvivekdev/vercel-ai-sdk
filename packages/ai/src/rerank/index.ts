export { rerank as experimental_rerank } from './rerank';
export type { RerankResult as Experimental_RerankResult } from './rerank-result';
