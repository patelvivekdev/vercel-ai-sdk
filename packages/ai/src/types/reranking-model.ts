import { RerankingModelV2, RerankingModelV2Result } from '@ai-sdk/provider';

/**
Reranking model that is used by the AI SDK Core functions.
*/
export type RerankingModel<VALUE> = RerankingModelV2<VALUE>;

/**Reranked index that contains the index of the document and its relevance score.
 */
export type RerankedResultIndex = RerankingModelV2Result;
