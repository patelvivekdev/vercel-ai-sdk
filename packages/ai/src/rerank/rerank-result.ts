import { RerankingModelUsage } from '../types/usage';
import { RerankedDocument } from '../types/reranking-model';

/**
The result of an `rerank` call.
It contains the documents, the reranked results, and additional information.
 */
export interface RerankResult<VALUE> {
  /**
  The documents that were reranked.
    */
  readonly documents: Array<VALUE>;

  /**
  The reranked documents is the list of object with the index, relevance score, and the document that have been reranked.
    */
  readonly rerankedDocuments: Array<RerankedDocument<VALUE>>;

  /**
  The reranking token usage.
    */
  readonly usage: RerankingModelUsage;

  /**
  Optional raw response data.
     */
  readonly response?: {
    /**
  Response headers.
       */
    headers?: Record<string, string>;

    /**
    The response body.
    */
    body?: unknown;
  };
}
