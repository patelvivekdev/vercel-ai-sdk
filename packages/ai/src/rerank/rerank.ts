import { RerankingModelV2Result } from '@ai-sdk/provider';
import { ProviderOptions } from '@ai-sdk/provider-utils';
import { prepareRetries } from '../../src/util/prepare-retries';
import { assembleOperationName } from '../telemetry/assemble-operation-name';
import { getBaseTelemetryAttributes } from '../telemetry/get-base-telemetry-attributes';
import { getTracer } from '../telemetry/get-tracer';
import { recordSpan } from '../telemetry/record-span';
import { selectTelemetryAttributes } from '../telemetry/select-telemetry-attributes';
import { TelemetrySettings } from '../telemetry/telemetry-settings';
import { RerankingModel } from '../types';
import { splitArray } from '../util/split-array';
import { RerankResult } from './rerank-result';
import { UnsupportedModelVersionError } from '../error';

/**
Rerank documents using an reranking model. The type of the value is defined by the reranking model.

@param model - The Reranking model to use.
@param values - The documents that should be reranking.
@param query - The query is a string that represents the query to rerank the documents against.
@param topK - Top k documents to rerank.
@param returnDocuments - Return the reranked documents in the response (In same order as indices).

@param maxRetries - Maximum number of retries. Set to 0 to disable retries. Default: 2.
@param abortSignal - An optional abort signal that can be used to cancel the call.
@param headers - Additional HTTP headers to be sent with the request. Only applicable for HTTP-based providers.

@returns A result object that contains the reranked documents, the reranked indices, and additional information.
 */
export async function rerank<VALUE>({
  model,
  values,
  query,
  topK,
  returnDocuments = false,
  maxParallelCalls = Infinity,
  maxRetries: maxRetriesArg,
  abortSignal,
  headers,
  providerOptions,
  experimental_telemetry: telemetry,
}: {
  /**
The reranking model to use.
  */
  model: RerankingModel<VALUE>;

  /**
The documents that should be reranked.
   */
  values: Array<VALUE>;

  /**
The query is a string that represents the query to rerank the documents against.
   */
  query: string;

  /**
Top k documents to rerank.
  */
  topK: number;

  /**
Return the reranked documents in the response (In same order as indices).

@default false
   */
  returnDocuments?: boolean;

  /**
Maximum number of retries per reranking model call. Set to 0 to disable retries.

@default 2
   */
  maxRetries?: number;

  /**
Abort signal.
 */
  abortSignal?: AbortSignal;

  /**
Additional headers to include in the request.
Only applicable for HTTP-based providers.
 */
  headers?: Record<string, string>;

  /**
   * Optional telemetry configuration (experimental).
   */
  experimental_telemetry?: TelemetrySettings;

  /**
    Additional provider-specific options. They are passed through
    to the provider from the AI SDK and enable provider-specific
    functionality that can be fully encapsulated in the provider.
    */
  providerOptions?: ProviderOptions;

  /**
   * Maximum number of concurrent requests.
   *
   * @default Infinity
   */
  maxParallelCalls?: number;
}): Promise<RerankResult<VALUE>> {
  if (model.specificationVersion !== 'v2') {
    throw new UnsupportedModelVersionError({
      version: model.specificationVersion,
      provider: model.provider,
      modelId: model.modelId,
    });
  }
  const { maxRetries, retry } = prepareRetries({ maxRetries: maxRetriesArg });

  const baseTelemetryAttributes = getBaseTelemetryAttributes({
    model,
    telemetry,
    headers,
    settings: { maxRetries },
  });

  const tracer = getTracer(telemetry);

  return recordSpan({
    name: 'ai.rerank',
    attributes: selectTelemetryAttributes({
      telemetry,
      attributes: {
        ...assembleOperationName({ operationId: 'ai.rerank', telemetry }),
        ...baseTelemetryAttributes,
        'ai.values': { input: () => values.map(v => JSON.stringify(v)) },
      },
    }),
    tracer,
    fn: async span => {
      const [maxDocumentsPerCall, supportsParallelCalls] = await Promise.all([
        model.maxDocumentsPerCall,
        model.supportsParallelCalls,
      ]);

      if (maxDocumentsPerCall == null || maxDocumentsPerCall === Infinity) {
        const { rerankedIndices, usage, rerankedDocuments, response } =
          await retry(() =>
            recordSpan({
              name: 'ai.rerank.doRerank',
              attributes: selectTelemetryAttributes({
                telemetry,
                attributes: {
                  ...assembleOperationName({
                    operationId: 'ai.rerank.doRerank',
                    telemetry,
                  }),
                  ...baseTelemetryAttributes,
                  // specific settings that only make sense on the outer level:
                  'ai.values': {
                    input: () => values.map(value => JSON.stringify(value)),
                  },
                },
              }),
              tracer,
              fn: async doRerankSpan => {
                const modelResponse = await model.doRerank({
                  values,
                  query,
                  topK,
                  returnDocuments,
                  providerOptions,
                  abortSignal,
                  headers,
                });

                const rerankedIndices = modelResponse.rerankedIndices;
                const usage = modelResponse.usage ?? { tokens: NaN };
                const rerankedDocuments = modelResponse.rerankedDocuments ?? [];

                doRerankSpan.setAttributes(
                  selectTelemetryAttributes({
                    telemetry,
                    attributes: {
                      'ai.rerankedIndices': {
                        output: () =>
                          rerankedIndices.map(index => JSON.stringify(index)),
                      },
                      'ai.usage.tokens': usage.tokens,
                      'ai.rerankedDocuments': {
                        output: () =>
                          rerankedDocuments.map(doc => JSON.stringify(doc)),
                      },
                    },
                  }),
                );

                return {
                  rerankedIndices,
                  usage,
                  rerankedDocuments,
                  response: modelResponse.response,
                };
              },
            }),
          );

        span.setAttributes(
          selectTelemetryAttributes({
            telemetry,
            attributes: {
              'ai.rerankedIndices': {
                output: () => JSON.stringify(rerankedIndices),
              },
              'ai.rerankedDocuments': {
                output: () => JSON.stringify(rerankedDocuments),
              },
              'ai.usage.tokens': usage.tokens,
            },
          }),
        );

        return new DefaultRerankResult({
          values,
          rerankedIndices,
          rerankedDocuments,
          usage,
          responses: [response],
        });
      }

      // split the values into chunks that are small enough for the model:
      const valueChunks = splitArray(values, maxDocumentsPerCall);

      const rerankedIndices: Array<RerankingModelV2Result> = [];
      const rerankedDocuments: Array<VALUE> = [];
      const responses: Array<
        | {
            headers?: Record<string, string>;
            body?: unknown;
          }
        | undefined
      > = [];
      let tokens = 0;

      const parallelChunks = splitArray(
        valueChunks,
        supportsParallelCalls ? maxParallelCalls : 1,
      );

      for (const parallelChunk of parallelChunks) {
        const results = await Promise.all(
          parallelChunk.map(chunk => {
            return retry(() => {
              return recordSpan({
                name: 'ai.rerank.doRerank',
                attributes: selectTelemetryAttributes({
                  telemetry,
                  attributes: {
                    ...assembleOperationName({
                      operationId: 'ai.rerank.doRerank',
                      telemetry,
                    }),
                    ...baseTelemetryAttributes,
                    'ai.values': {
                      input: () => chunk.map(value => JSON.stringify(value)),
                    },
                  },
                }),
                tracer,
                fn: async doRerankSpan => {
                  const modelResponse = await model.doRerank({
                    values: chunk,
                    query,
                    topK,
                    providerOptions,
                    returnDocuments,
                    abortSignal,
                    headers,
                  });

                  const chunkIndices = modelResponse.rerankedIndices;
                  const usage = modelResponse.usage ?? { tokens: NaN };
                  const chunkedRerankedDocuments =
                    modelResponse.rerankedDocuments ?? [];

                  doRerankSpan.setAttributes(
                    selectTelemetryAttributes({
                      telemetry,
                      attributes: {
                        'ai.rerankedIndices': {
                          output: () => JSON.stringify(chunkIndices),
                        },
                        'ai.usage.tokens': usage.tokens,
                        'ai.rerankedDocuments': {
                          output: () =>
                            JSON.stringify(chunkedRerankedDocuments),
                        },
                      },
                    }),
                  );

                  return {
                    rerankedIndices: chunkIndices,
                    usage,
                    rerankedDocuments: chunkedRerankedDocuments,
                  };
                },
              });
            });
          }),
        );

        for (const result of results) {
          rerankedIndices.push(...result.rerankedIndices);
          rerankedDocuments.push(...result.rerankedDocuments);
          tokens += result.usage.tokens;
        }
      }

      span.setAttributes(
        selectTelemetryAttributes({
          telemetry,
          attributes: {
            'ai.rerankedIndices': {
              output: () => JSON.stringify(rerankedIndices),
            },
            'ai.rerankedDocuments': {
              output: () => JSON.stringify(rerankedDocuments),
            },
            'ai.usage.tokens': tokens,
          },
        }),
      );

      return new DefaultRerankResult({
        values,
        rerankedIndices,
        rerankedDocuments,
        usage: { tokens },
        responses,
      });
    },
  });
}

class DefaultRerankResult<VALUE> implements RerankResult<VALUE> {
  readonly documents: RerankResult<VALUE>['documents'];
  readonly rerankedIndices: RerankResult<VALUE>['rerankedIndices'];
  readonly rerankedDocuments: RerankResult<VALUE>['documents'];
  readonly usage: RerankResult<VALUE>['usage'];
  readonly responses: RerankResult<VALUE>['responses'];

  constructor(options: {
    values: RerankResult<VALUE>['documents'];
    rerankedIndices: RerankResult<VALUE>['rerankedIndices'];
    rerankedDocuments: RerankResult<VALUE>['documents'];
    usage: RerankResult<VALUE>['usage'];
    responses?: RerankResult<VALUE>['responses'];
  }) {
    this.documents = options.values;
    this.rerankedIndices = options.rerankedIndices;
    this.rerankedDocuments = options.rerankedDocuments;
    this.usage = options.usage;
    this.responses = options.responses;
  }
}
