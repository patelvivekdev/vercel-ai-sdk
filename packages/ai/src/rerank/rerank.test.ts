import assert from 'node:assert';
import { MockTracer } from '../test/mock-tracer';
import { rerank } from './rerank';
import {
  mockRerank,
  MockRerankingModelV2,
} from '../test/mock-reranking-model-v2';
import { createResolvablePromise } from '../util/create-resolvable-promise';
import { RerankedResultIndex } from '../types';

const dummyResultIndices: Array<RerankedResultIndex> = [
  { index: 1, relevance_score: 0.9 },
  { index: 0, relevance_score: 0.8 },
  { index: 2, relevance_score: 0.7 },
];

const testDocuments = [
  'sunny day at the beach',
  'rainy day in the city',
  'cloudy day in the mountains',
];

const dummyRerankedDocuments = [
  'rainy day in the city',
  'sunny day at the beach',
  'cloudy day in the mountains',
];

const query = 'rainy day';

const topK = 3;

describe('model.supportsParallelCalls', () => {
  it('should not parallelize when false', async () => {
    const events: string[] = [];
    let callCount = 0;

    const resolvables = [
      createResolvablePromise<void>(),
      createResolvablePromise<void>(),
      createResolvablePromise<void>(),
    ];

    const rerankPromise = rerank({
      model: new MockRerankingModelV2({
        supportsParallelCalls: false,
        maxDocumentsPerCall: 1,
        doRerank: async () => {
          const index = callCount++;
          events.push(`start-${index}`);

          await resolvables[index].promise;
          events.push(`end-${index}`);

          return {
            rerankedIndices: [dummyResultIndices[index]],
            response: { headers: {}, body: {} },
          };
        },
      }),
      values: testDocuments,
      query,
      topK,
    });

    resolvables.forEach(resolvable => {
      resolvable.resolve();
    });

    const { rerankedIndices } = await rerankPromise;

    expect(events).toStrictEqual([
      'start-0',
      'end-0',
      'start-1',
      'end-1',
      'start-2',
      'end-2',
    ]);

    expect(rerankedIndices).toStrictEqual(dummyResultIndices);
  });

  it('should parallelize when true', async () => {
    const events: string[] = [];
    let callCount = 0;

    const resolvables = [
      createResolvablePromise<void>(),
      createResolvablePromise<void>(),
      createResolvablePromise<void>(),
    ];

    const rerankPromise = rerank({
      model: new MockRerankingModelV2({
        supportsParallelCalls: true,
        maxDocumentsPerCall: 1,
        doRerank: async () => {
          const index = callCount++;
          events.push(`start-${index}`);

          await resolvables[index].promise;
          events.push(`end-${index}`);

          return {
            rerankedIndices: [dummyResultIndices[index]],
            response: { headers: {}, body: {} },
          };
        },
      }),
      values: testDocuments,
      query,
      topK,
    });

    resolvables.forEach(resolvable => {
      resolvable.resolve();
    });

    const { rerankedIndices } = await rerankPromise;

    expect(events).toStrictEqual([
      'start-0',
      'start-1',
      'start-2',
      'end-0',
      'end-1',
      'end-2',
    ]);

    expect(rerankedIndices).toStrictEqual(dummyResultIndices);
  });

  it('should support maxParallelCalls', async () => {
    const events: string[] = [];
    let callCount = 0;

    const resolvables = [
      createResolvablePromise<void>(),
      createResolvablePromise<void>(),
      createResolvablePromise<void>(),
    ];

    const rerankPromise = rerank({
      maxParallelCalls: 2,
      model: new MockRerankingModelV2({
        supportsParallelCalls: true,
        maxDocumentsPerCall: 1,
        doRerank: async () => {
          const index = callCount++;
          events.push(`start-${index}`);

          await resolvables[index].promise;
          events.push(`end-${index}`);

          return {
            rerankedIndices: [dummyResultIndices[index]],
            response: { headers: {}, body: {} },
          };
        },
      }),
      values: testDocuments,
      query,
      topK,
    });

    resolvables.forEach(resolvable => {
      resolvable.resolve();
    });

    const { rerankedIndices } = await rerankPromise;

    expect(events).toStrictEqual([
      'start-0',
      'start-1',
      'end-0',
      'end-1',
      'start-2',
      'end-2',
    ]);

    expect(rerankedIndices).toStrictEqual(dummyResultIndices);
  });
});

describe('result.reranking', () => {
  it('should reranking documents', async () => {
    const result = await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 5,
        doRerank: mockRerank(
          testDocuments,
          dummyResultIndices,
          dummyRerankedDocuments,
        ),
      }),
      values: testDocuments,
      query,
      topK,
    });

    assert.deepStrictEqual(result.rerankedIndices, dummyResultIndices);
  });

  it('should reranking documents when several calls are required', async () => {
    let callCount = 0;
    const result = await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 2,
        doRerank: async ({ values }) => {
          switch (callCount++) {
            case 0:
              assert.deepStrictEqual(values, testDocuments.slice(0, 2));
              return {
                rerankedIndices: dummyResultIndices.slice(0, 2),
              };
            case 1:
              assert.deepStrictEqual(values, testDocuments.slice(2));
              return {
                rerankedIndices: dummyResultIndices.slice(2),
              };
            default:
              throw new Error('Unexpected call');
          }
        },
      }),
      values: testDocuments,
      query,
      topK,
    });

    assert.deepStrictEqual(result.rerankedIndices, dummyResultIndices);
  });
});

describe('result.responses', () => {
  it('should include responses in the result', async () => {
    let callCount = 0;
    const result = await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 1,
        doRerank: async ({ values }) => {
          switch (callCount++) {
            case 0:
              assert.deepStrictEqual(values, [testDocuments[0]]);
              return {
                rerankedIndices: dummyResultIndices.slice(0, 1),
                response: {
                  body: { first: true },
                },
              };
            case 1:
              assert.deepStrictEqual(values, [testDocuments[1]]);
              return {
                rerankedIndices: dummyResultIndices.slice(1, 2),
                response: {
                  body: { second: true },
                },
              };
            case 2:
              assert.deepStrictEqual(values, [testDocuments[2]]);
              return {
                rerankedIndices: dummyResultIndices.slice(2),
                response: {
                  body: { third: true },
                },
              };
            default:
              throw new Error('Unexpected call');
          }
        },
      }),
      values: testDocuments,
      query,
      topK,
    });

    expect(result.responses).toMatchSnapshot();
  });
});

describe('result.value', () => {
  it('should include value in the result', async () => {
    const result = await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 5,
        doRerank: mockRerank(
          testDocuments,
          dummyResultIndices,
          dummyRerankedDocuments,
        ),
      }),
      values: testDocuments,
      query,
      topK,
      returnDocuments: true,
    });

    assert.deepStrictEqual(result.rerankedDocuments, dummyRerankedDocuments);
  });

  it('should include documents in the result when returnDocuments is true', async () => {
    const result = await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 5,
        doRerank: mockRerank(
          testDocuments,
          dummyResultIndices,
          dummyRerankedDocuments,
        ),
      }),
      values: testDocuments,
      query,
      topK,
      returnDocuments: true,
    });

    assert.deepStrictEqual(result.rerankedIndices, dummyResultIndices);
    assert.deepStrictEqual(result.rerankedDocuments, dummyRerankedDocuments);
  });
});

describe('result.usage', () => {
  it('should include usage in the result', async () => {
    let callCount = 0;
    const result = await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 2,
        doRerank: async () => {
          switch (callCount++) {
            case 0:
              return {
                rerankedIndices: dummyResultIndices.slice(0, 2),
                usage: { tokens: 10 },
              };
            case 1:
              return {
                rerankedIndices: dummyResultIndices.slice(2),
                usage: { tokens: 20 },
              };
            default:
              throw new Error('Unexpected call');
          }
        },
      }),
      values: testDocuments,
      query,
      topK,
    });

    assert.deepStrictEqual(result.usage, { tokens: 30 });
  });
});

describe('options.headers', () => {
  it('should set headers', async () => {
    const result = await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 5,
        doRerank: async ({ headers }) => {
          assert.deepStrictEqual(headers, {
            'custom-request-header': 'request-header-value',
          });

          return { rerankedIndices: dummyResultIndices };
        },
      }),
      values: testDocuments,
      query,
      topK,
      headers: { 'custom-request-header': 'request-header-value' },
    });

    assert.deepStrictEqual(result.rerankedIndices, dummyResultIndices);
  });
});

describe('options.providerOptions', () => {
  it('should pass provider options to model', async () => {
    const model = new MockRerankingModelV2({
      maxDocumentsPerCall: 5,
      doRerank: async ({ providerOptions }) => {
        return {
          rerankedIndices: [dummyResultIndices[0]],
        };
      },
    });

    vi.spyOn(model, 'doRerank');

    await rerank({
      model,
      values: ['test-input'],
      providerOptions: {
        aProvider: { someKey: 'someValue' },
      },
      query: 'test-query',
      topK: 1,
    });

    expect(model.doRerank).toHaveBeenCalledWith({
      abortSignal: undefined,
      headers: undefined,
      providerOptions: {
        aProvider: { someKey: 'someValue' },
      },
      values: ['test-input'],
      query: 'test-query',
      returnDocuments: false,
      topK: 1,
    });
  });
});

describe('telemetry', () => {
  let tracer: MockTracer;

  beforeEach(() => {
    tracer = new MockTracer();
  });

  it('should not record any telemetry data when not explicitly enabled', async () => {
    await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 5,
        doRerank: mockRerank(
          testDocuments,
          dummyResultIndices,
          dummyRerankedDocuments,
        ),
      }),
      values: testDocuments,
      query,
      topK,
    });
    assert.deepStrictEqual(tracer.jsonSpans, []);
  });

  it('should record telemetry data when enabled (single call path)', async () => {
    await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: null,
        doRerank: mockRerank(
          testDocuments,
          dummyResultIndices,
          dummyRerankedDocuments,
          {
            tokens: 10,
          },
        ),
      }),
      values: testDocuments,
      query,
      topK,
      experimental_telemetry: {
        isEnabled: true,
        functionId: 'test-function-id',
        metadata: {
          test1: 'value1',
          test2: false,
        },
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });

  it('should record telemetry data when enabled (multiple calls path)', async () => {
    let callCount = 0;
    await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: 2,
        doRerank: async ({ values }) => {
          switch (callCount++) {
            case 0:
              assert.deepStrictEqual(values, testDocuments.slice(0, 2));
              return {
                rerankedIndices: dummyResultIndices.slice(0, 2),
                usage: { tokens: 20 },
              };
            case 1:
              assert.deepStrictEqual(values, testDocuments.slice(2));
              return {
                rerankedIndices: dummyResultIndices.slice(2),
                usage: { tokens: 10 },
              };
            default:
              throw new Error('Unexpected call');
          }
        },
      }),
      values: testDocuments,
      query,
      topK,
      experimental_telemetry: {
        isEnabled: true,
        functionId: 'test-function-id',
        metadata: {
          test1: 'value1',
          test2: false,
        },
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });

  it('should not record telemetry inputs / outputs when disabled', async () => {
    await rerank({
      model: new MockRerankingModelV2({
        maxDocumentsPerCall: null,
        doRerank: mockRerank(
          testDocuments,
          dummyResultIndices,
          dummyRerankedDocuments,
          {
            tokens: 10,
          },
        ),
      }),
      values: testDocuments,
      query,
      topK,
      experimental_telemetry: {
        isEnabled: true,
        recordInputs: false,
        recordOutputs: false,
        tracer,
      },
    });

    expect(tracer.jsonSpans).toMatchSnapshot();
  });
});
