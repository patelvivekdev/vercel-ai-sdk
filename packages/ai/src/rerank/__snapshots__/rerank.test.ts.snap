// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`result.responses > should include responses in the result 1`] = `[]`;

exports[`telemetry > should not record telemetry inputs / outputs when disabled 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank",
      "ai.settings.maxRetries": 2,
      "ai.usage.tokens": 10,
      "operation.name": "ai.rerank",
    },
    "events": [],
    "name": "ai.rerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.settings.maxRetries": 2,
      "ai.usage.tokens": 10,
      "operation.name": "ai.rerank.doRerank",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
]
`;

exports[`telemetry > should record telemetry data when enabled (multiple calls path) 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank",
      "ai.rerankedDocuments": "[]",
      "ai.rerankedIndices": "[{"index":1,"relevance_score":0.9},{"index":0,"relevance_score":0.8},{"index":2,"relevance_score":0.7}]",
      "ai.settings.maxRetries": 2,
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 30,
      "ai.values": [
        ""sunny day at the beach"",
        ""rainy day in the city"",
        ""cloudy day in the mountains"",
      ],
      "operation.name": "ai.rerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.rerankedDocuments": "[]",
      "ai.rerankedIndices": "[{"index":1,"relevance_score":0.9},{"index":0,"relevance_score":0.8}]",
      "ai.settings.maxRetries": 2,
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 20,
      "ai.values": [
        ""sunny day at the beach"",
        ""rainy day in the city"",
      ],
      "operation.name": "ai.rerank.doRerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.rerankedDocuments": "[]",
      "ai.rerankedIndices": "[{"index":2,"relevance_score":0.7}]",
      "ai.settings.maxRetries": 2,
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 10,
      "ai.values": [
        ""cloudy day in the mountains"",
      ],
      "operation.name": "ai.rerank.doRerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
]
`;

exports[`telemetry > should record telemetry data when enabled (single call path) 1`] = `
[
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank",
      "ai.rerankedDocuments": "["rainy day in the city","sunny day at the beach","cloudy day in the mountains"]",
      "ai.rerankedIndices": "[{"index":1,"relevance_score":0.9},{"index":0,"relevance_score":0.8},{"index":2,"relevance_score":0.7}]",
      "ai.settings.maxRetries": 2,
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 10,
      "ai.values": [
        ""sunny day at the beach"",
        ""rainy day in the city"",
        ""cloudy day in the mountains"",
      ],
      "operation.name": "ai.rerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank",
  },
  {
    "attributes": {
      "ai.model.id": "mock-model-id",
      "ai.model.provider": "mock-provider",
      "ai.operationId": "ai.rerank.doRerank",
      "ai.rerankedDocuments": [
        ""rainy day in the city"",
        ""sunny day at the beach"",
        ""cloudy day in the mountains"",
      ],
      "ai.rerankedIndices": [
        "{"index":1,"relevance_score":0.9}",
        "{"index":0,"relevance_score":0.8}",
        "{"index":2,"relevance_score":0.7}",
      ],
      "ai.settings.maxRetries": 2,
      "ai.telemetry.functionId": "test-function-id",
      "ai.telemetry.metadata.test1": "value1",
      "ai.telemetry.metadata.test2": false,
      "ai.usage.tokens": 10,
      "ai.values": [
        ""sunny day at the beach"",
        ""rainy day in the city"",
        ""cloudy day in the mountains"",
      ],
      "operation.name": "ai.rerank.doRerank test-function-id",
      "resource.name": "test-function-id",
    },
    "events": [],
    "name": "ai.rerank.doRerank",
  },
]
`;
